packages/assets-controllers/src/NftController.ts

```diff
@@ -944,6 +991,145 @@ export class NftController extends BaseController<NftConfig, NftState> {
     });
   }
 
+  async validateWatchNftAndFetchMetadata(
+    asset: NftAsset,
+    type: NFTStandardType,
+    accountAddress: string,
+  ) {
+    const { chainId } = this.config;
+
+    const { address: contractAddress, tokenId } = asset;
+
+    // Validate parameters
+    if (!type) {
+      throw ethErrors.rpc.invalidParams('Asset type is required');
+    }
+
+    if (type !== ERC721 && type !== ERC1155) {
+      throw ethErrors.rpc.invalidParams(
+        `Non NFT asset type ${type} not supported by watchNft`,
+      );
+    }
+
+    if (!contractAddress || !tokenId) {
+      throw ethErrors.rpc.invalidParams(
+        'Both address and tokenId are required',
+      );
+    }
+
+    if (!isAddress(contractAddress)) {
+      throw ethErrors.rpc.invalidParams('Invalid address');
+    }
+
+    if (!/^\d+$/u.test(tokenId)) {
+      throw ethErrors.rpc.invalidParams('Invalid tokenId');
+    }
+
+    // Check if the suggested NFT is already pending
+    // if (
+    // TODO: check if the suggested NFT is already pending in approval controller state
+    // ) {
+    //   throw ethErrors.rpc.internal('Suggested NFT already pending');
+    // }
+
+    // Check if the suggested NFT is already being watched by the selected account
+    if (
+      this.state.allNfts?.[accountAddress]?.[chainId]?.find(
+        ({ address: watchedAssetAddress, tokenId: watchedAssetTokenId }) => {
+          return (
+            watchedAssetAddress === contractAddress &&
+            watchedAssetTokenId === tokenId
+          );
+        },
+      )
+    ) {
+      throw ethErrors.rpc.internal(
```

Comment:
dont think `internal` makes sense here for these NFT specific errors, they could be their own error code?

Reply:
yeah I wasn't sure what the appropriate error codes were here. I can/should create new ones?